File: get_endereco.py
#function to show data from a Brazilian address when user input a CEP and getting a response from 'viacep.com.br'
def get_endereco(request, **kwargs):
    import json
    endereco = json.loads(kwargs.get('str'))
		#cidade = city
    cidade = models.Cidade.objects.filter(cod_ibge=endereco['ibge']).get()
		#bairro = neighborhood
    bairros = models.Bairro.objects.filter(cidade=cidade)
    bairros_nome = list(models.Bairro.objects.filter(cidade=cidade).values_list('nome', flat=True))
    bairro_json = endereco['bairro'] if endereco.get('bairro') else ''
    bairro_id = int()

		#verify if the response has 'bairro' info
    if bairro_json:
        #verify if 'bairro' doesn't exist in DB
        if not bairro_json.lower() in [nome.lower() for nome in bairros_nome]:
            #if 'bairro' doesn't exist, create it
            lista_bairros = bairros.order_by('bairro')
            new_cod_bairro = int(list(lista_bairros)[len(lista_bairros)-1].bairro + 1)
            new_nome_bairro = bairro_json
            new_cidade_bairro = cidade
            new_empresa_bairro = utils.get_empresa_logada(request)
            
            #new 'bairro' instance
            new_bairro = models.Bairro(bairro=new_cod_bairro, nome=new_nome_bairro, cidade=new_cidade_bairro, empresa=new_empresa_bairro)
            #save it
            new_bairro.save()
            
            #update the 'bairro' search to add the new one
            bairros = models.Bairro.objects.filter(cidade=cidade)
    
        #data from 'bairro' that was informed via CEP
        bairro_cep = bairros.filter(nome=bairro_json).get()

		#create the json response
    response = {
        'endereco' : endereco['logradouro'] if endereco.get('logradouro') else '',
        'cidade_id' : cidade.pk,
        'cidade' : f'{cidade.cod_ibge} - {cidade.nome}, {cidade.uf}',
        'bairros_lista' : [
            {
                'cod': bairro.pk,
                'desc': f'{bairro.bairro} - {bairro.nome}'
            }
            for bairro in bairros
        ],
        'bairro_id' : bairro_cep.pk if bairro_json else '',
        'bairro' : f'{bairro_cep.bairro} - {bairro_cep.nome}' if bairro_json else ''
    }
		return JsonResponse(response, safe=False, json_dumps_params={'ensure_ascii': False})

File: get_endereco.js
//functions to set a Brazilian address when CEP is informed (front-end)

function setAddress(event){
	var cep = document.getElementById("id_cep").value;
	//verify if CEP was typed
	if (cep != ''){
	//verify if it's only numbers
	if (/^\d+$/.test(cep)){
		//verify if it has 8 digits
		if (cep.length == 8) {
			//set the request for 'viacep.com.br' to get address info
			let request = new XMLHttpRequest();
			let method = "GET";
			let url = 'https://viacep.com.br/ws/' + cep + '/json/';
			request.open(method, url);
			request.onload = function() {
				var endjson = request.response;
				var parsedJSON = JSON.parse(endjson);
				if (!parsedJSON.erro) {
					if (parsedJSON.complemento) {
						if (parsedJSON.complemento.indexOf("/") > -1) {
							parsedJSON.complemento = parsedJSON.complemento.replace("/", "-");
						}
					}
	
					endjson = JSON.stringify(parsedJSON);
	
					//set the request for the Django back-end for other infos
					let newrequest = new XMLHttpRequest();
					let newmethod = "GET";
					let newurl = document.getElementById("cep_div").getAttribute("data-url");
					newurlcut = newurl.slice(0, newurl.indexOf('k'))+endjson;
					newrequest.open(newmethod, newurlcut);
					newrequest.onload = function(){
						var newendereco = JSON.parse(newrequest.response);
						//select the informed city via CEP
						document.getElementById("id_endereco").value = newendereco.endereco;
						document.getElementById("select2-id_cidade-container").innerHTML = newendereco.cidade;
						document.getElementById("id_cidade").value = newendereco.cidade_id;
						//remake the neighborhood list on html (fuction bellow)
						setBairro(newendereco.bairros_lista);
						//select the informed neighborhood via CEP
						document.getElementById("select2-id_bairro-container").innerHTML = newendereco.bairro;
						document.getElementById("id_bairro").value = newendereco.bairro_id;
					};
					newrequest.send();
				} else {
					//show error (fuction bellow)
					exibirErro('ERRO! CEP não existe! Favor verificar!');
				}
			};
			request.send();
		} else {
			//show error (function bellow)
			exibirErro('ERRO! O número precisa ter 8 dígitos (Exemplo: 00000000)')
		}
		} else {
			//show error (function bellow)
			exibirErro('ERRO! Digite apenas NÚMEROS (Exemplo: 00000000)');
		}
	}
}

//change the HTML list of neighborhoods
function setBairro(bairros){
	var select_bairro = document.getElementById("id_bairro");
	var v = "";
	
	for (var i = 0; i < bairros.length; i++) {
		v += "<option value='" + bairros[i]["cod"] + "'>" + bairros[i]["desc"] + "</option>"
	}
	select_bairro.innerHTML = v;
}

//show error div
function exibirErro(mensagem) {
	var errorDiv = document.getElementById("errorDiv");
	errorDiv.innerText = mensagem;
	errorDiv.classList.add("alert", "alert-danger", "alert-dismissible");
	errorDiv.style.display = "block";

	var contentDiv = document.getElementById("content-div");
	contentDiv.parentNode.insertBefore(errorDiv, contentDiv);

	setTimeout(function() {
		errorDiv.style.display = "none";
	}, 5000);
}
